Logging and reporting

Julien00859 opened the pull request:
The pull request embed multiple changes, the most important is the
introduction of logging capacities and a /report route to report
problematic snippets. Those are the two first commits.

Next we have some fixes and cleanups.

---

**feat: new /report endpoint to report a snippet**

There have been incident report about users using the software to host
malwares. While those malware cannot be executed nor on the server nor
in the web page, users might copy/paste them in their own environment
and run them.

In order to take rapid action to remove those snippets, a new "report"
button have been included in the top left corner of the highlight.html
page so users can report a problematic snippet.

Closes #110

---

**feat: introduce logging**

We got feedback from system administrators willing get logs whenever a
snippet was created. Introducing logging was not as simple as it should
be because it needs to correctly hooks to the server software logger
when running behind a wsgi server.

Because we needed much more logics in the config module, we decided to
enclose the setup in a dedicated function in order to limit the exported
tokens.

---

**fix: identify form using name**

---

**fix: lang and extension parsing**

Fine tunning of 5868f01, I am a fucking monkey.

---

**doc: the page description links the github repo**

Julien00859 committed:
fix: lang and extension parsing

Fine tunning of 5868f01, I am a fucking monkey.

Julien00859 commented on bin/assets/styles/style.css:
```diff
@@ -82,7 +82,7 @@ body {
 }
 
-.controls {
+div.controls {
```
arrêter de pisser tu veux, connard, sombre merde, LIS TES PUTAIN DE DIFF AVANT DE COMMIT

Julien00859 commented on bin/views/head.html:
```diff
@@ -5,7 +5,7 @@
 <meta name=author content="ReadTheDocs">
 <meta name=title content="ReadTheDocs Bin">
-<meta name=description content="A simple pastebin">
+<meta name=description content="https://github.com/readthedocs-fr/bin-server">
```
Qu'est-ce que ça fou dans cette PR ?

Julien00859 replied:
Héhé petit commit à part en furt :eyes:

Mesteery replied:
Mais du coup pourquoi ce changement ?

Julien00859 commented on bin/assets/scripts/report.js:
```diff
@@ -0,0 +1,10 @@
+	].join('\n'));
+	if (name === null) return false;
+    reportForm.name = name;
+});
```
Rigolo ce code de merde, surtout qu'il marche pas

Mesteery replied:
Il est moche ton code, tu ne respecte même pas le code style habituel...
Et aussi c'est pas comme ça que ça marche pas...

Julien00859 replied:
Comment il faudrait faire ??

Julien00859 commented on bin/config.py:
```diff
@@ -20,11 +20,14 @@
 """
 
 
+import logging
```
encore faudrait-il qu'il log qqch MERCI GUNICORN DE MERDE

Julien00859 commented on bin/controller.py:
```diff
@@ -124,6 +126,7 @@ def post_new():
         raise bt.HTTPError(400, str(exc))
 
     snippet = models.Snippet.create(code, maxusage, lifetime, parentid)
+    logger.info("New %s snippet: %s", lang, snippet.id)
```
Rigolo, voila 2 mois que j'ai ça en prod, oui je modifie le code source en prod comme un crado

Julien00859 reviewed (commented):
IL EST PRESQUE 4H ET TU AS FAIT QUE ÇA MAIS WTF

Julien00859 committed:
fix: identify form using name

Julien00859 committed:
feat: introduce logging

We got feedback from system administrators willing get logs whenever a
snippet was created. Introducing logging was not as simple as it should
be because it needs to correctly hooks to the server software logger
when running behind a wsgi server.

Because we needed much more logics in the config module, we decided to
enclose the setup in a dedicated function in order to limit the exported
tokens.

Julien00859 committed:
feat: new /report endpoint to report a snippet

There have been incident report about users using the software to host
malwares. While those malware cannot be executed nor on the server nor
in the web page, users might copy/paste them in their own environment
and run them.

In order to take rapid action to remove those snippets, a new "report"
button have been included in the top left corner of the highlight.html
page so users can report a problematic snippet.

Closes #110

Julien00859 committed:
doc: the page description links the github repo

Mesteery commented on bin/assets/scripts/report.js:
```diff
@@ -0,0 +1,11 @@
+const promptMessage = (
+      "Vous allez signaler ce snippet auprès de l'administrateur.\n"
```
Pas besoin de parenthèses. De plus, renseigne toi sur les *template literals*

Julien00859 replied:
Je connais et c'est pas beau

Mesteery replied:
Si ça l'est..., bien plus que cette horreur

Mesteery commented on bin/assets/scripts/report.js:
```diff
@@ -0,0 +1,11 @@
+    + "Tout abus peut mener à la coupure du service.\n\n"
+    + "Veuillez entrer votre nom ou votre adresse email:"
+);
+const reportForm = document.forms['report'];
```
```suggestion
const reportForm = document.forms.report;

```

Mesteery commented on bin/assets/scripts/report.js:
```diff
@@ -0,0 +1,11 @@
+);
+const reportForm = document.forms['report'];
+reportForm.addEventListener('submit', (event) => {
+    const name = prompt(promptMessage).trim();
```
Il faut vérifier si l'utilisateur à entrer quelque chose et que s'il entré quelque chose que la longeur  soit > 0 (ou autre limite)

Julien00859 replied:
si l'utilisateur est en noscript, alors il ne pourra de toute façon rien rentrer, j'ai préféré rester cohérant et les laisser ne pas mettre de nom (puis ça plantera niveau backend)

Mesteery replied:
Oui mais là ton code peut ne pas fonctionner (prompt peut renvoyer undefined) et puis rien ne coûte à vérifier la longueur avant d'envoyer

Julien00859 replied:
puis, cf mon message en dessous, ça ne submit pas quand le champ est laissé vide

Mesteery replied:
Bah si, et ça laissera une erreur : `undefined.lenght`...

Mesteery commented on bin/assets/scripts/report.js:
```diff
@@ -0,0 +1,11 @@
+reportForm.addEventListener('submit', (event) => {
+    const name = prompt(promptMessage).trim();
+    reportForm.name.value = name;
+    return name.length;
```
return inutile

Julien00859 replied:
dans un submit il est utile, la fonction doit renvoyer true ou false si il faut propager le submit ou pas

Mesteery replied:
Il se propagera sans problème tant qu'on `preventDefault` pas

Julien00859 replied:
Dingue ça, ça a changé depuis la dernière fois que j'ai fait du JS, je corrige ça

Mesteery commented on bin/assets/styles/report.css:
```diff
@@ -0,0 +1,10 @@
+form#report {
```
Pour être consistent avec le reste enleve `form` et `.5` -> `0.5`.

Sinon je comprend pas pourquoi tu position absolute et display none input

Julien00859 replied:
Je trouve ça dommage qu'on ne laisse que l'id, tu perds l'info sur quel type d'élément tu es en train de styliser

Mesteery replied:
Bah bof, à part avoir Alzheimer je pense qu'on est capable de quoi s'agit-il, surtout que le form est inutile et presque redondant

Mesteery commented on bin/config.py:
```diff
@@ -20,31 +23,73 @@
-REDIS_PORT = int(os.getenv('REDIS_PORT', 6379))
-REDIS_DB = int(os.getenv('REDIS_DB', 0))
+
+def _setup():
```
Pourquoi faire une fonction ?
Sinon pourquoi préfixer avec un _ ?

Julien00859 replied:
C'est marqué dans le message de commit

Mesteery commented on bin/views/highlight.html:
```diff
@@ -2,11 +2,25 @@
 <html lang=fr>
     <head>
         % include('head.html')
+        <link rel=stylesheet href="/assets/styles/report.css">
         <link rel=stylesheet href="/assets/styles/monokai.css">
         <script src="/assets/scripts/loc.js" defer></script>
+        <script src="/assets/scripts/report.js" defer></script>
```
Est-vraiment nécessaire d'ajouter des fichiers supplémentaires à charger ? Je pense que `loc.js` et `report.js` pourraient fusionnés et le fichier final pourrait être renommé.
*Vivement #55*

Mesteery commented on bin/assets/styles/report.css:
```diff
@@ -0,0 +1,10 @@
+	position: absolute;
+	top: 0.5em;
+	right: 0.5em;
+	width: 1em;
```
```suggestion
	width: 1.5em;
```
c'était un peu trop petit

Mesteery commented on bin/views/highlight.html:
```diff
@@ -2,11 +2,25 @@
     <body>
 {{!codehl}}
+
+        <form id=report name=report action="/report" method=POST>
```
```suggestion
        <form id=report title="Signaler" name=report action="/report" method=POST>
```

Mesteery commented on bin/assets/styles/report.css:
```diff
@@ -0,0 +1,10 @@
+	right: 0.5em;
+	width: 1em;
+}
+
```
```suggestion

#report label {
	cursor: pointer;
}

```

NoeMastrorillo commented on bin/assets/scripts/report.js:
```diff
@@ -0,0 +1,17 @@
+    return;
+  }
+
+  reportForm.name.value = name;
```
```suggestion
  reportForm.name.value = name;
  alert("The snippet have been reported.");
```

NoeMastrorillo commented on bin/controller.py:
```diff
@@ -176,3 +179,29 @@ def get_raw(snippetid, ext=None):
+        raise bt.HTTPError(404, "Snippet not found")
+    logger.warning("The snippet %s got reported by %s", snippetid, name)
+
+    return bt.HTTPResponse("The snippet have been reported.")
```
```suggestion
    bt.redirect('/')
```
ou
```suggestion
    bt.redirect(f'/{snippetid}')
```
(Il faudrait dans ce deuxième cas gérer l'extension via le formulaire).

NoeMastrorillo reviewed (commented):
Je trouve ça un peu brouillon de simplement return une `HTTPResponse` à la fin du report.
Je verrais à la place un `alert("The snippet have been reported.")` côté js et ensuite un redirect sur `/` ou sur le snippet.
